scalyr_agent/builtin_monitors/windows_event_log_monitor.py

```diff
@@ -773,6 +801,91 @@ def log_event(self, event):
         finally:
             self._bookmark_lock.release()
 
+    def _param_placeholder_value(self, channel, provider, param):
+        if not re.match("^%%[0-9]+$", param):
+            return param
+
+        if channel == "Security":
+            provider = "Security"
+
+        param_key = "%s-%s-%s" % (channel, provider, param)
+        if param_key in self._param_cache:
+            return self._param_cache[param_key]
+
+        dll_key = "%s-%s" % (channel, provider)
+        if dll_key in self._dll_cache:
+            if not isinstance(self._dll_cache[dll_key], _DLL):
+                return param
+        else:
+            # Before opening a new handle to a dll check if another handle to the same dll is open;
+            # multiple event log providers may ultimately reference the same dll.
+
+            try:
+                dllpath = _DLL.dllpath(channel, provider)
+            except Exception as e:
+                self._dll_cache[dll_key] = e
+                return param
+
+            found = False
+            for dll in self._dll_cache.values():
+                if dll.path == dllpath:
+                    self._dll_cache[dll_key] = dll
+                    found = True
+                    break
+
+            if not found:
+                try:
+                    self._dll_cache[dll_key] = _DLL(channel, provider)
```

Comment:
Does this mean we keep an open file handle to each loaded DLL which is referenced?

I guess that should not be a big issue since I see the cache size is limited to 10 entries.

Reply:
Yes